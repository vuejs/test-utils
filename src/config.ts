import { ComponentPublicInstance } from 'vue'
import { GlobalMountOptions } from './types'
import { VueWrapper } from './vueWrapper'
import { DOMWrapper } from './domWrapper'

interface GlobalConfigOptions {
  global: Required<GlobalMountOptions>
  plugins: {
    VueWrapper: Pluggable<VueWrapper<ComponentPublicInstance>>
    DOMWrapper: Pluggable<DOMWrapper<Element>>
  }
  renderStubDefaultSlot: boolean
}

interface Plugin<Instance, O> {
  handler(instance: Instance): Record<string, any>
  handler(instance: Instance, options: O): Record<string, any>
  options: O
}

class Pluggable<Instance = DOMWrapper<Element>> {
  installedPlugins: Plugin<Instance, any>[] = []

  install<O>(handler: (instance: Instance) => Record<string, any>): void
  install<O>(
    handler: (instance: Instance, options: O) => Record<string, any>,
    options: O
  ): void
  install<O>(
    handler: (instance: Instance, options?: O) => Record<string, any>,
    options?: O
  ): void {
    if (typeof handler !== 'function') {
      console.error('plugin.install must receive a function')
      handler = () => ({})
    }
    this.installedPlugins.push({ handler, options })
  }

  extend(instance: Instance) {
    const invokeSetup = ({ handler, options }: Plugin<Instance, any>) => {
      return handler(instance, options) // invoke the setup method passed to install
    }
    const bindProperty = ([property, value]: [string, any]) => {
      ;(instance as any)[property] =
        typeof value === 'function' ? value.bind(instance) : value
    }
    const addAllPropertiesFromSetup = (setupResult: Record<string, any>) => {
      setupResult = typeof setupResult === 'object' ? setupResult : {}
      Object.entries(setupResult).forEach(bindProperty)
    }

    this.installedPlugins.map(invokeSetup).forEach(addAllPropertiesFromSetup)
  }

  /** For testing */
  reset() {
    this.installedPlugins = []
  }
}

export const config: GlobalConfigOptions = {
  global: {
    stubs: {
      transition: true,
      'transition-group': true
    },
    provide: {},
    components: {},
    config: {},
    directives: {},
    mixins: [],
    mocks: {},
    plugins: [],
    renderStubDefaultSlot: false
  },
  plugins: {
    VueWrapper: new Pluggable(),
    DOMWrapper: new Pluggable()
  },
  renderStubDefaultSlot: false
}
