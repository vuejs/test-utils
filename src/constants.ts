export const MOUNT_ELEMENT_ID = 'app'
