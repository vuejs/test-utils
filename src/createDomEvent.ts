// @ts-ignore todo - No DefinitelyTyped package exists for this
import eventTypes from 'dom-event-types'
import { Key } from 'readline'

const keyCodesByKeyName = {
  backspace: 8,
  tab: 9,
  enter: 13,
  esc: 27,
  space: 32,
  pageup: 33,
  pagedown: 34,
  end: 35,
  home: 36,
  left: 37,
  up: 38,
  right: 39,
  down: 40,
  insert: 45,
  delete: 46
} as const

// modifiers to keep an eye on
const ignorableKeyModifiers = ['stop', 'prevent', 'self', 'exact']
const systemKeyModifiers = ['ctrl', 'shift', 'alt', 'meta'] as const
const mouseKeyModifiers = ['left', 'middle', 'right'] as const

type KeyName = keyof typeof keyCodesByKeyName
type Modifier =
  | typeof systemKeyModifiers[number]
  | typeof mouseKeyModifiers[number]

interface TriggerOptions {
  code?: String
  key?: String
  keyCode?: Number
  [custom: string]: any
}

interface EventParams {
  eventType: string
  modifiers: KeyName[]
  options?: TriggerOptions
}

/**
 * Groups modifiers into lists
 */
function generateModifiers(modifiers: KeyName[], isOnClick: boolean) {
  const keyModifiers: KeyName[] = []
  const systemModifiers: Modifier[] = []

  for (let i = 0; i < modifiers.length; i++) {
    const modifier: KeyName | Modifier = modifiers[i]

    // addEventListener() options, e.g. .passive & .capture, that we dont need to handle
    if (ignorableKeyModifiers.includes(modifier)) {
      continue
    }
    // modifiers that require special conversion
    // if passed a left/right key modifier with onClick, add it here as well.
    if (
      systemKeyModifiers.includes(
        modifier as Exclude<typeof modifier, KeyName>
      ) ||
      (mouseKeyModifiers.includes(
        modifier as Exclude<typeof modifier, KeyName>
      ) &&
        isOnClick)
    ) {
      systemModifiers.push(modifier as Modifier)
    } else {
      keyModifiers.push(modifier)
    }
  }

  return {
    keyModifiers,
    systemModifiers
  }
}

function getEventProperties(eventParams: EventParams) {
  let { modifiers, options = {}, eventType } = eventParams

  let isOnClick = eventType === 'click'

  const { keyModifiers, systemModifiers } = generateModifiers(
    modifiers,
    isOnClick
  )

  if (isOnClick) {
    // if it's a right click, it should fire a `contextmenu` event
    if (systemModifiers.includes('right')) {
      eventType = 'contextmenu'
      options.button = 2
      // if its a middle click, fire a `mouseup` event
    } else if (systemModifiers.includes('middle')) {
      eventType = 'mouseup'
      options.button = 1
    }
  }

  const meta = eventTypes[eventType] || {
    eventInterface: 'Event',
    cancelable: true,
    bubbles: true
  }

  // convert `shift, ctrl` to `shiftKey, ctrlKey`
  // allows trigger('keydown.shift.ctrl.n') directly
  const systemModifiersMeta = systemModifiers.reduce(
    (all: Record<string, boolean>, key) => {
      all[`${key}Key`] = true
      return all
    },
    {}
  )

  // get the keyCode for backwards compat
  const keyCode =
    keyCodesByKeyName[keyModifiers[0]] ||
    (options && (options.keyCode || options.code))

  const eventProperties = {
    ...systemModifiersMeta, // shiftKey, metaKey etc
    ...options, // What the user passed in as the second argument to #trigger
    bubbles: meta.bubbles,
    cancelable: meta.cancelable,
    // Any derived options should go here
    keyCode,
    code: keyCode,
    // if we have a `key`, use it, otherwise dont set anything (allows user to pass custom key)
    ...(keyModifiers[0] ? { key: keyModifiers[0] } : {})
  }

  return {
    eventProperties,
    meta,
    eventType
  }
}

function createEvent(eventParams: EventParams) {
  const { eventProperties, meta, eventType } = getEventProperties(eventParams)

  // user defined eventInterface
  const metaEventInterface = window[meta.eventInterface]

  const SupportedEventInterface =
    typeof metaEventInterface === 'function' ? metaEventInterface : window.Event

  return new SupportedEventInterface(
    eventType,
    // event properties can only be added when the event is instantiated
    // custom properties must be added after the event has been instantiated
    eventProperties
  )
}

function createDOMEvent(eventString: String, options?: TriggerOptions) {
  // split eventString like `keydown.ctrl.shift.c` into `keydown` and array of modifiers
  const [eventType, ...modifiers] = eventString.split('.')

  const eventParams: EventParams = {
    eventType,
    modifiers: modifiers as KeyName[],
    options
  }
  const event: Event & TriggerOptions = createEvent(eventParams)
  const eventPrototype = Object.getPrototypeOf(event)

  // attach custom options to the event, like `relatedTarget` and so on.
  options &&
    Object.keys(options).forEach((key) => {
      const propertyDescriptor = Object.getOwnPropertyDescriptor(
        eventPrototype,
        key
      )
      const canSetProperty = !(
        propertyDescriptor && propertyDescriptor.set === undefined
      )
      if (canSetProperty) {
        event[key] = options[key]
      }
    })
  return event
}

export { TriggerOptions, createDOMEvent, keyCodesByKeyName, KeyName }
