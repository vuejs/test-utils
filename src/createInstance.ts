import {
  h,
  createApp,
  defineComponent,
  reactive,
  shallowReactive,
  ref,
  AppConfig,
  ComponentOptions,
  ConcreteComponent,
  DefineComponent,
  transformVNodeArgs
} from 'vue'

import { MountingOptions, Slot } from './types'
import {
  getComponentsFromStubs,
  getDirectivesFromStubs,
  isFunctionalComponent,
  isObject,
  isObjectComponent,
  isScriptSetup,
  mergeGlobalProperties
} from './utils'
import { processSlot } from './utils/compileSlots'
import { attachEmitListener } from './emit'
import { registerStub } from './stubs'
import {
  isLegacyFunctionalComponent,
  unwrapLegacyVueExtendComponent
} from './utils/vueCompatSupport'
import { createVNodeTransformer } from './vnodeTransformers/util'
import {
  createStubComponentsTransformer,
  CreateStubComponentsTransformerConfig
} from './vnodeTransformers/stubComponentsTransformer'
import { createStubDirectivesTransformer } from './vnodeTransformers/stubDirectivesTransformer'
import { isDeepRef } from './utils/isDeepRef'

const MOUNT_OPTIONS: ReadonlyArray<keyof MountingOptions<any>> = [
  'attachTo',
  'attrs',
  'data',
  'props',
  'slots',
  'global',
  'shallow'
] as const

function getInstanceOptions(
  options: MountingOptions<any> & Record<string, any>
): Record<string, any> {
  if (options.methods) {
    console.warn(
      "Passing a `methods` option to mount was deprecated on Vue Test Utils v1, and it won't have any effect on v2. For additional info: https://vue-test-utils.vuejs.org/upgrading-to-v1/#setmethods-and-mountingoptions-methods"
    )
    delete options.methods
  }

  const resultOptions = { ...options }
  for (const key of Object.keys(options)) {
    if (MOUNT_OPTIONS.includes(key as keyof MountingOptions<any>)) {
      delete resultOptions[key]
    }
  }
  return resultOptions
}

// implementation
export function createInstance(
  inputComponent: DefineComponent<{}, {}, any, any, any, any>,
  options?: MountingOptions<any> & Record<string, any>
) {
  // normalize the incoming component
  const originalComponent = unwrapLegacyVueExtendComponent(inputComponent)
  let component: ConcreteComponent
  const instanceOptions = getInstanceOptions(options ?? {})

  const rootComponents: CreateStubComponentsTransformerConfig['rootComponents'] =
    {}

  if (
    isFunctionalComponent(originalComponent) ||
    isLegacyFunctionalComponent(originalComponent)
  ) {
    component = defineComponent({
      compatConfig: {
        MODE: 3,
        INSTANCE_LISTENERS: false,
        INSTANCE_ATTRS_CLASS_STYLE: false,
        COMPONENT_FUNCTIONAL: isLegacyFunctionalComponent(originalComponent)
          ? 'suppress-warning'
          : false
      },
      props: originalComponent.props || {},
      setup:
        (props, { attrs, slots }) =>
        () =>
          h(originalComponent, { ...props, ...attrs }, slots),
      ...instanceOptions
    })
    rootComponents.functional = originalComponent
  } else if (isObjectComponent(originalComponent)) {
    component = { ...originalComponent, ...instanceOptions }
  } else {
    component = originalComponent
  }

  rootComponents.component = component
  // We've just replaced our component with its copy
  // Let's register it as a stub so user can find it
  registerStub({ source: originalComponent, stub: component })

  function slotToFunction(slot: Slot) {
    switch (typeof slot) {
      case 'function':
        return slot
      case 'object':
        return () => h(slot)
      case 'string':
        return processSlot(slot)
      default:
        throw Error(`Invalid slot received.`)
    }
  }

  // handle any slots passed via mounting options
  const slots =
    options?.slots &&
    Object.entries(options.slots).reduce(
      (
        acc: { [key: string]: Function },
        [name, slot]: [string, Slot]
      ): { [key: string]: Function } => {
        if (Array.isArray(slot)) {
          const normalized = slot.map(slotToFunction)
          acc[name] = (args: unknown) => normalized.map((f) => f(args))
          return acc
        }

        acc[name] = slotToFunction(slot)
        return acc
      },
      {}
    )

  // override component data with mounting options data
  if (options?.data) {
    const providedData = options.data()
    if (isObjectComponent(originalComponent)) {
      // component is guaranteed to be the same type as originalComponent
      const objectComponent = component as ComponentOptions
      const originalDataFn = originalComponent.data || (() => ({}))
      objectComponent.data = (vm) => ({
        ...originalDataFn.call(vm, vm),
        ...providedData
      })
    } else {
      throw new Error(
        'data() option is not supported on functional and class components'
      )
    }
  }

  const MOUNT_COMPONENT_REF = 'VTU_COMPONENT'
  // we define props as reactive so that way when we update them with `setProps`
  // Vue's reactivity system will cause a rerender.
  const refs = shallowReactive<Record<string, unknown>>({})
  const props = reactive<Record<string, unknown>>({})

  Object.entries({
    ...options?.attrs,
    ...options?.propsData,
    ...options?.props,
    ref: MOUNT_COMPONENT_REF
  }).forEach(([k, v]) => {
    if (isDeepRef(v)) {
      refs[k] = v
    } else {
      props[k] = v
    }
  })

  const global = mergeGlobalProperties(options?.global)
  if (isObjectComponent(component)) {
    component.components = { ...component.components, ...global.components }
  }

  const componentRef = ref(null)
  // create the wrapper component
  const Parent = defineComponent({
    name: 'VTU_ROOT',
    setup() {
      return {
        [MOUNT_COMPONENT_REF]: componentRef
      }
    },
    render() {
      return h(component as ComponentOptions, { ...props, ...refs }, slots)
    }
  })

  // create the app
  const app = createApp(Parent)

  // add tracking for emitted events
  // this must be done after `createApp`: https://github.com/vuejs/test-utils/issues/436
  attachEmitListener()

  // global mocks mixin
  if (global?.mocks) {
    const mixin: ComponentOptions = {
      beforeCreate() {
        // we need to differentiate components that are or not not `script setup`
        // otherwise we run into a proxy set error
        // due to https://github.com/vuejs/core/commit/f73925d76a76ee259749b8b48cb68895f539a00f#diff-ea4d1ddabb7e22e17e80ada458eef70679af4005df2a1a6b73418fec897603ceR404
        // introduced in Vue v3.2.45
        // Also ensures not to include option API components in this block
        // since they can also have setup state but need to be patched using
        // the regular method.
        if (isScriptSetup(this)) {
          // add the mocks to setupState
          for (const [k, v] of Object.entries(
            global.mocks as { [key: string]: any }
          )) {
            // we do this in a try/catch, as some properties might be read-only
            try {
              this.$.setupState[k] = v
              // eslint-disable-next-line no-empty
            } catch (e) {}
          }
          // also intercept the proxy calls to make the mocks available on the instance
          // (useful when a template access a global function like $t and the developer wants to mock it)
          // eslint-disable-next-line no-extra-semi
          ;(this.$ as any).proxy = new Proxy((this.$ as any).proxy, {
            get(target, key) {
              if (key in global.mocks) {
                return global.mocks[key as string]
              }
              return target[key]
            }
          })
        } else {
          for (const [k, v] of Object.entries(
            global.mocks as { [key: string]: any }
          )) {
            // eslint-disable-next-line no-extra-semi
            ;(this as any)[k] = v
          }
        }
      }
    }

    app.mixin(mixin)
  }

  // AppConfig
  if (global.config) {
    for (const [k, v] of Object.entries(global.config) as [
      keyof Omit<AppConfig, 'isNativeTag'>,
      any
    ][]) {
      app.config[k] = isObject(app.config[k])
        ? Object.assign(app.config[k]!, v)
        : v
    }
  }

  // use and plugins from mounting options
  if (global.plugins) {
    for (const plugin of global.plugins) {
      if (Array.isArray(plugin)) {
        app.use(plugin[0], ...plugin.slice(1))
        continue
      }
      app.use(plugin)
    }
  }

  // use any mixins from mounting options
  if (global.mixins) {
    for (const mixin of global.mixins) app.mixin(mixin)
  }

  if (global.components) {
    for (const key of Object.keys(global.components)) {
      // avoid registering components that are stubbed twice
      if (!(key in global.stubs)) {
        app.component(key, global.components[key])
      }
    }
  }

  if (global.directives) {
    for (const key of Object.keys(global.directives))
      app.directive(key, global.directives[key])
  }

  // provide any values passed via provides mounting option
  if (global.provide) {
    for (const key of Reflect.ownKeys(global.provide)) {
      // @ts-ignore: https://github.com/microsoft/TypeScript/issues/1863
      app.provide(key, global.provide[key])
    }
  }

  // stubs
  // even if we are using `mount`, we will still
  // stub out Transition and Transition Group by default.
  transformVNodeArgs(
    createVNodeTransformer({
      rootComponents,
      transformers: [
        createStubComponentsTransformer({
          rootComponents,
          stubs: getComponentsFromStubs(global.stubs),
          shallow: options?.shallow,
          renderStubDefaultSlot: global.renderStubDefaultSlot
        }),
        createStubDirectivesTransformer({
          directives: getDirectivesFromStubs(global.stubs)
        })
      ]
    })
  )

  // users expect stubs to work with globally registered
  // components so we register stubs as global components to avoid
  // warning about not being able to resolve component
  //
  // component implementation provided here will never be called
  // but we need name to make sure that stubComponents will
  // properly stub this later by matching stub name
  //
  // ref: https://github.com/vuejs/test-utils/issues/249
  // ref: https://github.com/vuejs/test-utils/issues/425
  if (global?.stubs) {
    for (const name of Object.keys(getComponentsFromStubs(global.stubs))) {
      if (!app.component(name)) {
        app.component(name, { name })
      }
    }
  }

  return {
    app,
    props,
    componentRef
  }
}
