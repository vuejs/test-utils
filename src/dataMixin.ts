import { getCurrentInstance } from 'vue'

export const createDataMixin = (data: Record<string, unknown>) => {
  return {
    created() {
      for (const [k, v] of Object.entries(data)) {
        getCurrentInstance()!.data = { ...getCurrentInstance()!.data, [k]: v }
      }
    }
  }
}
