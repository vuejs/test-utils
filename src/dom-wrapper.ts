import { nextTick } from 'vue'

import { WrapperAPI } from './types'
import { ErrorWrapper } from './error-wrapper'

export class DOMWrapper<ElementType extends Element> implements WrapperAPI {
  element: ElementType

  constructor(element: ElementType) {
    this.element = element
  }

  classes(className?) {
    const classes = this.element.classList

    if (className) return classes.contains(className)

    return Array.from(classes)
  }

  attributes(key?: string) {
    const attributes = this.element.attributes
    const attributeMap = {}
    for (let i = 0; i < attributes.length; i++) {
      const att = attributes.item(i)
      attributeMap[att.localName] = att.value
    }

    return key ? attributeMap[key] : attributeMap
  }

  exists() {
    return true
  }

  text() {
    return this.element.textContent?.trim()
  }

  html() {
    return this.element.outerHTML
  }

  find<T extends Element>(selector: string): DOMWrapper<T> | ErrorWrapper {
    const result = this.element.querySelector<T>(selector)
    if (result) {
      return new DOMWrapper<T>(result)
    }

    return new ErrorWrapper({ selector })
  }

  findAll<T extends Element>(selector: string): DOMWrapper<T>[] {
    return Array.from(this.element.querySelectorAll<T>(selector)).map(
      (x) => new DOMWrapper(x)
    )
  }

  async setChecked(checked: boolean = true) {
    // typecast so we get typesafety
    const element = (this.element as unknown) as HTMLInputElement

    if (element.tagName !== 'INPUT') {
      throw Error(
        `You need to call setChecked on an input element. You called it on a ${this.element.tagName}`
      )
    }

    // we do not want to trigger an event if the user
    // attempting set the same value twice
    // this is beacuse in a browser setting checked = true when it is
    // already true is a no-op; no change event is triggered
    if (checked === element.checked) {
      return
    }

    element.checked = checked
    return this.trigger('change')
  }

  async trigger(eventString: string) {
    const evt = document.createEvent('Event')
    evt.initEvent(eventString)

    if (this.element) {
      this.element.dispatchEvent(evt)
      return nextTick
    }
  }
}
