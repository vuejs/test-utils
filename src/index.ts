import { DOMWrapper } from './domWrapper'
import { VueWrapper } from './vueWrapper'
import BaseWrapper from './baseWrapper'
import { mount, shallowMount } from './mount'
import { MountingOptions } from './types'
import { RouterLinkStub } from './components/RouterLinkStub'
import { createWrapperError } from './errorWrapper'
import { config } from './config'
import { flushPromises } from './utils/flushPromises'
import { enableAutoUnmount, disableAutoUnmount } from './utils/autoUnmount'

export {
  mount,
  shallowMount,
  enableAutoUnmount,
  disableAutoUnmount,
  RouterLinkStub,
  VueWrapper,
  DOMWrapper,
  BaseWrapper,
  config,
  flushPromises,
  MountingOptions,
  createWrapperError
}
