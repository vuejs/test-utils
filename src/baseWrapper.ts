import { textContent } from './utils'
import { createDOMEvent } from './createDomEvent'
import type { TriggerOptions } from './createDomEvent'
import { nextTick } from 'vue'

export default class BaseWrapper<ElementType extends Element> {
  private readonly wrapperElement: ElementType

  get element() {
    return this.wrapperElement
  }

  constructor(element: ElementType) {
    this.wrapperElement = element
  }

  classes(): string[]
  classes(className: string): boolean
  classes(className?: string): string[] | boolean {
    const classes = this.element.classList

    if (className) return classes.contains(className)

    return Array.from(classes)
  }

  attributes(): { [key: string]: string }
  attributes(key: string): string
  attributes(key?: string): { [key: string]: string } | string {
    const attributes = Array.from(this.element.attributes)
    const attributeMap: Record<string, string> = {}
    for (const attribute of attributes) {
      attributeMap[attribute.localName] = attribute.value
    }

    return key ? attributeMap[key] : attributeMap
  }

  text() {
    return textContent(this.element)
  }

  exists() {
    return true
  }

  async trigger(eventString: string, options?: TriggerOptions) {
    if (options && options['target']) {
      throw Error(
        `[vue-test-utils]: you cannot set the target value of an event. See the notes section ` +
          `of the docs for more detailsâ€”` +
          `https://vue-test-utils.vuejs.org/api/wrapper/trigger.html`
      )
    }

    const isDisabled = () => {
      const validTagsToBeDisabled = [
        'BUTTON',
        'COMMAND',
        'FIELDSET',
        'KEYGEN',
        'OPTGROUP',
        'OPTION',
        'SELECT',
        'TEXTAREA',
        'INPUT'
      ]
      const hasDisabledAttribute = this.attributes().disabled !== undefined
      const elementCanBeDisabled = validTagsToBeDisabled.includes(
        this.element.tagName
      )

      return hasDisabledAttribute && elementCanBeDisabled
    }

    if (this.element && !isDisabled()) {
      const event = createDOMEvent(eventString, options)
      this.element.dispatchEvent(event)
    }

    return nextTick()
  }
}
