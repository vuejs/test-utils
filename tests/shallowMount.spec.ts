import { defineAsyncComponent, defineComponent } from 'vue'
import { mount, shallowMount, VueWrapper } from '../src'
import ComponentWithChildren from './components/ComponentWithChildren.vue'
import ScriptSetupWithChildren from './components/ScriptSetupWithChildren.vue'

describe('shallowMount', () => {
  it('renders props for stubbed component in a snapshot', () => {
    expect.addSnapshotSerializer({
      test(wrapper: VueWrapper<any>) {
        return '__app' in wrapper
      },
      serialize(wrapper: VueWrapper<any>) {
        return wrapper.html()
      }
    })

    const MyLabel = defineComponent({
      props: ['val'],
      template: '<label :for="val">{{ val }}</label>'
    })

    const AsyncComponent = defineAsyncComponent(async () => ({
      name: 'AsyncComponentName',
      template: '<span>AsyncComponent</span>'
    }))

    const Component = defineComponent({
      components: { MyLabel, AsyncComponent },
      template: '<div><MyLabel val="username" /><AsyncComponent /></div>',
      data() {
        return {
          foo: 'bar'
        }
      }
    })

    const wrapper = shallowMount(Component)

    expect(wrapper.html()).toBe(
      '<div>\n' +
        '  <my-label-stub val="username"></my-label-stub>\n' +
        '  <async-component-stub></async-component-stub>\n' +
        '</div>'
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('stubs all components automatically using { shallow: true }', () => {
    const wrapper = mount(ComponentWithChildren, { shallow: true })
    expect(wrapper.html()).toEqual(
      '<div class="ComponentWithChildren">\n' +
        '  <hello-stub></hello-stub>\n' +
        '  <component-with-input-stub></component-with-input-stub>\n' +
        '  <component-without-name-stub></component-without-name-stub>\n' +
        '  <script-setup-stub></script-setup-stub>\n' +
        '  <with-props-stub></with-props-stub>\n' +
        '</div>'
    )
  })

  it('stubs all components automatically using shallowMount', () => {
    const wrapper = shallowMount(ComponentWithChildren)
    expect(wrapper.html()).toEqual(
      '<div class="ComponentWithChildren">\n' +
        '  <hello-stub></hello-stub>\n' +
        '  <component-with-input-stub></component-with-input-stub>\n' +
        '  <component-without-name-stub></component-without-name-stub>\n' +
        '  <script-setup-stub></script-setup-stub>\n' +
        '  <with-props-stub></with-props-stub>\n' +
        '</div>'
    )
  })

  it('correctly renders slot content', () => {
    const ComponentWithSlot = defineComponent({
      template: '<div><slot></slot></div>'
    })

    const wrapper = shallowMount(ComponentWithSlot, {
      slots: {
        default: '<span class="slot-content">test</span>'
      }
    })
    expect(wrapper.find('.slot-content').exists()).toBe(true)
  })

  it('correctly stubs components inside slot', () => {
    const ComponentWithSlot = defineComponent({
      template: '<div><slot></slot></div>'
    })

    const Foo = defineComponent({
      name: 'Foo',
      template: '<div class="unstubbed-foo">OK</div>'
    })

    const wrapper = shallowMount(ComponentWithSlot, {
      global: {
        components: { Foo }
      },
      slots: {
        default: '<Foo />'
      }
    })

    expect(wrapper.find('.unstubbed-foo').exists()).toBe(false)
  })

  it('stubs all components, but allows providing custom stub', () => {
    const wrapper = mount(ComponentWithChildren, {
      shallow: true,
      global: {
        stubs: {
          Hello: { template: '<div>Override</div>' }
        }
      }
    })
    expect(wrapper.html()).toEqual(
      '<div class="ComponentWithChildren">\n' +
        '  <div>Override</div>\n' +
        '  <component-with-input-stub></component-with-input-stub>\n' +
        '  <component-without-name-stub></component-without-name-stub>\n' +
        '  <script-setup-stub></script-setup-stub>\n' +
        '  <with-props-stub></with-props-stub>\n' +
        '</div>'
    )
  })

  it('stubs all components, but allows disabling stub by passing false', () => {
    const wrapper = mount(ComponentWithChildren, {
      shallow: true,
      global: {
        stubs: {
          Hello: false
        }
      }
    })
    expect(wrapper.html()).toEqual(
      '<div class="ComponentWithChildren">\n' +
        '  <div id="root">\n' +
        '    <div id="msg">Hello world</div>\n' +
        '  </div>\n' +
        '  <component-with-input-stub></component-with-input-stub>\n' +
        '  <component-without-name-stub></component-without-name-stub>\n' +
        '  <script-setup-stub></script-setup-stub>\n' +
        '  <with-props-stub></with-props-stub>\n' +
        '</div>'
    )
  })

  it('stubs all components in a script setup component', () => {
    const wrapper = mount(ScriptSetupWithChildren, {
      shallow: true,
      global: {
        stubs: {
          Hello: { template: '<div>Override</div>' }
        }
      }
    })
    expect(wrapper.html()).toEqual(
      '<div>Override</div>\n' +
        '<component-with-input-stub></component-with-input-stub>\n' +
        '<anonymous-stub></anonymous-stub>\n' +
        '<anonymous-stub></anonymous-stub>\n' +
        '<with-props-stub></with-props-stub>'
    )
  })

  it('should render stubs correctly', () => {
    const ComponentToMount = defineComponent({
      template: `<div>
        <test-component class='component 1' />
        <test-component class='component 2' />
      </div>`
    })

    const TestComponent = defineComponent({
      template: `<div>
        this is just a test component
      </div>`
    })

    const wrapper = shallowMount(ComponentToMount, {
      global: {
        components: {
          TestComponent
        }
      }
    })

    expect(wrapper.html()).toBe(
      '<div>\n' +
        '  <test-component-stub class="component 1"></test-component-stub>\n' +
        '  <test-component-stub class="component 2"></test-component-stub>\n' +
        '</div>'
    )
  })
})
