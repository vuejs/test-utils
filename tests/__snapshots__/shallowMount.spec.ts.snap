// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallowMount renders props for stubbed component in a snapshot 1`] = `
<div>
  <my-label-stub val="username"></my-label-stub>
  <async-component-stub></async-component-stub>
</div>
`;
